servers/rendering/rendering_device_driver.h

```diff
@@ -0,0 +1,619 @@
+		uint32_t buffer_row_length = 0;
+		uint32_t buffer_texture_height = 0;
+		TextureSubresourceLayers texture_subresources;
+		Vector3i texture_offset;
+		Vector3i texture_region_size;
+	};
+
+	enum TextureCopyMode {
+		TEXTURE_COPY_MODE_COPY,
+		TEXTURE_COPY_MODE_RESOLVE,
+	};
+
+	virtual void command_clear_buffer(CommandBufferID p_cmd_buffer, BufferID p_buffer, uint64_t p_offset, uint64_t p_size) = 0;
+	virtual void command_copy_buffer(CommandBufferID p_cmd_buffer, BufferID p_src_buffer, BufferID p_dst_buffer, VectorView<BufferCopyRegion> p_regions) = 0;
+	virtual void command_copy_texture(CommandBufferID p_cmd_buffer, TextureID p_src_texture, TextureLayout p_src_texture_layout, TextureID p_dst_texture, TextureLayout p_dst_texture_layout, VectorView<TextureCopyRegion> p_regions, TextureCopyMode p_mode) = 0;
+
+	virtual void command_clear_color_texture(CommandBufferID p_cmd_buffer, TextureID p_texture, TextureLayout p_texture_layout, const Color &p_color, const TextureSubresourceRange &p_subresources) = 0;
+
+	virtual void command_copy_buffer_to_texture(CommandBufferID p_cmd_buffer, BufferID p_src_buffer, TextureID p_dst_texture, TextureLayout p_dst_texture_layout, VectorView<BufferTextureCopyRegion> p_regions) = 0;
+	virtual void command_copy_texture_to_buffer(CommandBufferID p_cmd_buffer, TextureID p_src_texture, TextureLayout p_src_texture_layout, BufferID p_dst_buffer, VectorView<BufferTextureCopyRegion> p_regions) = 0;
+
+	/******************/
+	/**** PIPELINE ****/
+	/******************/
+
+	virtual void pipeline_free(PipelineID p_pipeline_id) = 0;
+
+	// ----- BINDING -----
+
+	virtual void command_bind_push_constants(CommandBufferID p_cmd_buffer, ShaderID p_shader, uint32_t p_offset, VectorView<uint32_t> p_data) = 0;
+
+	// ----- CACHE -----
+
+	virtual bool pipeline_cache_create(const Vector<uint8_t> &p_data) = 0;
+	virtual void pipeline_cache_free() = 0;
+	virtual size_t pipeline_cache_query_size() = 0;
+	virtual Vector<Vector<uint8_t>> pipeline_cache_serialize() = 0;
+
+	/*******************/
+	/**** RENDERING ****/
+	/*******************/
+
+	// ----- SUBPASS -----
+
+	enum AttachmentLoadOp {
+		ATTACHMENT_LOAD_OP_LOAD = 0,
+		ATTACHMENT_LOAD_OP_CLEAR = 1,
+		ATTACHMENT_LOAD_OP_DONT_CARE = 2,
+		ATTACHMENT_LOAD_OP_NONE = 1000400000,
+	};
+
+	enum AttachmentStoreOp {
+		ATTACHMENT_STORE_OP_STORE = 0,
+		ATTACHMENT_STORE_OP_DONT_CARE = 1,
+		ATTACHMENT_STORE_OP_NONE = 1000301000,
+	};
+
+	struct Attachment {
+		DataFormat format = DATA_FORMAT_MAX;
+		TextureSamples samples = TEXTURE_SAMPLES_MAX;
+		AttachmentLoadOp load_op = ATTACHMENT_LOAD_OP_LOAD;
+		AttachmentStoreOp store_op = ATTACHMENT_STORE_OP_STORE;
+		AttachmentLoadOp stencil_load_op = ATTACHMENT_LOAD_OP_LOAD;
+		AttachmentStoreOp stencil_store_op = ATTACHMENT_STORE_OP_STORE;
+		TextureLayout initial_layout = TEXTURE_LAYOUT_UNDEFINED;
+		TextureLayout final_layout = TEXTURE_LAYOUT_UNDEFINED;
+	};
+
+	struct AttachmentReference {
+		static const uint32_t UNUSED = 0xffffffff;
```

Comment:
This constant seems like it'd be better off defined at the global level, as the code below uses this same constant but without a name.

Reply:
IMO this should go away as we remove subpasses anyway.

Reply:
The one below is for subpass, and is just a garbage, but fixed default value to make it apparent when it's not properly filled. This one is a proper value that can be specified as an (unused) attachment index.